pkg/dataobj/metastore/object.go

```diff
@@ -0,0 +1,357 @@
+		})
+	}
+
+	return streams, nil
+}
+
+func (m *ObjectMetastore) DataObjects(ctx context.Context, start, end time.Time, matchers ...*labels.Matcher) ([]DataObjectMeta, error) {
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all metastore paths for the time range
+	var storePaths []string
+	for path := range iterStorePaths(tenantID, start, end) {
+		storePaths = append(storePaths, path)
+	}
+
+	// List objects from all stores concurrently
+	paths, err := m.listObjectsFromStores(ctx, storePaths, start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	objects := make([]DataObjectMeta, 0, len(paths))
+	for _, path := range paths {
+		objects = append(objects, DataObjectMeta{Path: path})
+	}
+	return objects, nil
+}
+
+func (m *ObjectMetastore) Labels(ctx context.Context, start, end time.Time, matchers ...*labels.Matcher) ([]string, error) {
+	return nil, nil
+}
+
+func (m *ObjectMetastore) Values(ctx context.Context, start, end time.Time, matchers ...*labels.Matcher) ([]string, error) {
+	return nil, nil
+}
+
+func predicateFromMatchers(start, end time.Time, matchers ...*labels.Matcher) dataobj.StreamsPredicate {
+	if len(matchers) == 0 {
+		return nil
+	}
+
+	predicates := make([]dataobj.StreamsPredicate, 0, len(matchers)+1)
+	predicates = append(predicates, dataobj.TimeRangePredicate[dataobj.StreamsPredicate]{
+		StartTime:    start,
+		EndTime:      end,
+		IncludeStart: true,
+		IncludeEnd:   true,
+	})
+	for _, matcher := range matchers {
+		switch matcher.Type {
+		case labels.MatchEqual:
+			predicates = append(predicates, dataobj.LabelMatcherPredicate{
+				Name:  matcher.Name,
+				Value: matcher.Value,
+			})
+		case labels.MatchNotEqual:
+			predicates = append(predicates, dataobj.NotPredicate[dataobj.StreamsPredicate]{
+				Inner: dataobj.LabelMatcherPredicate{
+					Name:  matcher.Name,
+					Value: matcher.Value,
+				},
+			})
+		case labels.MatchRegexp:
+			predicates = append(predicates, dataobj.LabelFilterPredicate{
+				Name: matcher.Name,
+				Keep: func(name, value string) bool {
+					return matcher.Matches(value)
+				},
+			})
+		case labels.MatchNotRegexp:
+			predicates = append(predicates, dataobj.NotPredicate[dataobj.StreamsPredicate]{
+				Inner: dataobj.LabelFilterPredicate{
+					Name: matcher.Name,
+					Keep: func(name, value string) bool {
+						return !matcher.Matches(value)
+					},
+				},
+			})
+		}
+	}
+
+	if len(predicates) == 1 {
+		return predicates[0]
+	}
+
+	current := dataobj.AndPredicate[dataobj.StreamsPredicate]{
+		Left: predicates[0],
+	}
+
+	for _, predicate := range predicates[1:] {
+		and := dataobj.AndPredicate[dataobj.StreamsPredicate]{
+			Left:  predicate,
+			Right: current,
+		}
+		current = and
+	}
+	return current
+}
+
+// listObjectsFromStores concurrently lists objects from multiple metastore files
+func (m *ObjectMetastore) listObjectsFromStores(ctx context.Context, storePaths []string, start, end time.Time) ([]string, error) {
```

Comment:
Almost everything beyond this point moved from metastore.go and is used from the dataobj query path.